src/test/e2e/fixtures/auth.setup.js

```diff
@@ -0,0 +1,150 @@
+      }
+
+      if (!loginCounts[testFilePath]) {
+        loginCounts[testFilePath] = 0;
+      }
+      loginCounts[testFilePath]++;
+
+      try {
+        fs.writeFileSync(countsFilePath, JSON.stringify(loginCounts, null, 2), 'utf8');
+      } catch (writeError) {
+        console.error('Error writing loginCounts.json:', writeError);
+        continue; // Retry
+      }
+
+      fs.rmdirSync(lockDirPath);
+      return; // Success
+
+    } catch (error) {
+      if (error.code === 'EEXIST') {
+        await new Promise(resolve => setTimeout(resolve, retryDelay));
+      } else {
+        console.error('Error acquiring or releasing lock:', error);
+        throw error;
+      }
+    }
+  }
+  console.error(`Failed to acquire lock after ${maxRetries} attempts.`);
+  throw new Error(`Failed to acquire lock after ${maxRetries} attempts.`);
+}
+
+const roleCredentials = {
+  admin: {
+    username: process.env.OAUTH_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  viewer: {
+    username: process.env.OAUTH_VIEWER_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  superAdmin: {
+    username: process.env.OAUTH_SUPER_USER,
+    password: process.env.OAUTH_USER_PASSWORD
+  },
+  noRole: { username: process.env.OAUTH_TEST_USER_NO_ROLE, password: process.env.OAUTH_USER_PASSWORD }
+};
+
+exports.test = base.extend({
+  adminPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'admin', testInfo);
+    await use(page);
+  },
+  viewerPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'viewer', testInfo);
+    await use(page);
+  },
+  superAdminPage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'superAdmin', testInfo);
+    await use(page);
+  },
+  noRolePage: async ({ page }, use, testInfo) => {
+    await loginWithRole(page, 'noRole', testInfo);
+    await use(page);
+  },
+  page: async ({ page }, use) => {
+    await use(page);
+  }
+});
+
+async function loginWithRole(page, role, testInfo) {
+  const loginPage = new LoginPage(page);
+  const credentials = roleCredentials[role];
+  const testFilePath = testInfo.file;
+  const loginKey = role;
+
+  if (!hasLoggedIn[loginKey]) {
+    await loginPage.goto();
+    await expect(page).toHaveURL(/.*idam-web-public.*/);
+    console.log(`Attempting login for role: ${role}, user: ${credentials.username}`);
+    await loginPage.login(credentials.username, credentials.password);
+
+    const errorLocator = page.locator('.error-summary');
+    if (await errorLocator.isVisible()) {
+      const errorMessage = await errorLocator.textContent();
+      console.error(`Login failed for role ${role}: ${errorMessage}`);
+      throw new Error(`Login failed for role ${role}: ${errorMessage}`);
+    }
+    console.log(`Login successful for role: ${role}`);
+
+    await updateCounts(testFilePath);
+    hasLoggedIn[loginKey] = true;
+    await page.waitForTimeout(500);
```

Comment:
Does this need to be hardcoded 500?

Is there a way to get it to detect automatically? await page.waitForLoadState('networkidle') for example

Reply:
So, i've tried various of these - networkidle works sometimes, but not always if it's doing DOM refeshes via ajax calls. I was planning to do a big audit of all the various types of waits at the end so i don't spend too long overoptimising until it's finished though.